### PR TITLE
add missing tests

### PR DESCRIPTION
**Pull request by FCanberk**
Added test cases for the functions that does not have tests yet in util.__init__.py

### CODE REVIEW

File: src/odemis/util/test/util_test.py [hunk 1]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+                                                                "the closest index to 5 in %s" % in_exp)
+
+
+class NormalizeRectTestCase(unittest.TestCase):
+
+    def test_already_normalized(self):
+        """ Test normalizing a rectangle with already normalized values """
+        rect = (1, 2, 3, 4)
+        normalized_rect = util.normalize_rect(rect)
+        self.assertEqual(normalized_rect, (1, 2, 3, 4), "Failed to normalize the rectangle")
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:506]: You could replace  `(1, 2, 3, 4)` with `rect`, that makes it clear that the output is the same as the input

File: src/odemis/util/test/util_test.py [hunk 2]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+
+
+class ExpandRectTestCase(unittest.TestCase):
+
+    def test_expand_rect(self):
+        """ Test expanding a rectangle """
+        rect = (5, 3, 10, 7)
+        expanded_rect = util.expand_rect(util.normalize_rect(rect), 5)
+        expected_rect = (0, -2, 15, 12)
+        self.assertEqual(expanded_rect, expected_rect,  "Failed to get correct output for rectangle %s with "
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:541]: There's 2 spaces between `expected_rect,` and `"Failed...`

File: src/odemis/util/test/util_test.py [hunk 3]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+
+class FindPlotContentTestCase(unittest.TestCase):
+
+    def test_find_plot_content(self):
+        """ Test finding left and right most non-zero values"""
+        xd = (10, 15, 20, 25, 30)
+        yd = (0, 0, 1, -2, 0)
+        plot_content = util.find_plot_content(xd, yd)
+        self.assertEqual(plot_content, (20, 25), "Failed to get correct output while finding plot content "
+                                                 "for %s with values  %s %s" % (plot_content, xd, yd))
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:553]: I think it would also make sense to add a test where there is one or more zero in between the left and right most non-zero value. So something like this for xd and yd:
```
        xd = (10, 15, 16, 17, 20, 25, 30)
        yd = (0, 0, 1, 0, 0, -2, 0)
```

Comment (tepals) [reply src/odemis/util/test/util_test.py:553]: or should yd also be sorted?

File: src/odemis/util/test/util_test.py [hunk 4]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+
+
+class WrapToMpiPpiTestCase(unittest.TestCase):
+
+    def test_positive_pi(self):
+        """ Test value is smaller than +pi value """
+        converted_angle = util.wrap_to_mpi_ppi(4.65)
+        self.assertLessEqual(converted_angle, math.pi, "Failed to convert the angle to a value between -pi and +pi "
+                                                  "with the converted value: %s" % converted_angle)
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:562]: I'm not sure if it's a PEP8 rule or not, but it looks a bit nicer if you align the string with the string above it:
```
        self.assertLessEqual(converted_angle, math.pi, "Failed to convert the angle to a value between -pi and +pi "
                                                       "with the converted value: %s" % converted_angle)
```

File: src/odemis/util/test/util_test.py [hunk 5]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+
+    def test_negative_pi(self):
+        """ Test value is bigger than -pi value """
+        converted_angle = util.wrap_to_mpi_ppi(-5.1)
+        self.assertGreaterEqual(converted_angle, -math.pi, "Failed to convert the angle to a value between -pi and +pi "
+                                                      "with the converted value: %s" % converted_angle)
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:568]: same here, it looks a bit nicer if you align the string with the string above it

File: src/odemis/util/test/util_test.py [hunk 6]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+
+        new_dict = util.recursive_dict_update(dict1, dict2)
+        self.assertEqual(new_dict, expected_dict, "Failed to get correct output while updating the dictionary "
+                                                  " %s" % new_dict)
+
+
+class PairwiseTestCase(unittest.TestCase):
+
+    def test_pairwise(self):
+        """ Test pairing """
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:601]: I would say  `""" Test pairing two items in a tuple """`

File: src/odemis/util/test/util_test.py [hunk 7]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+        test = util.pairwise(("item 1", "item 2"))
+        test_list = list(test)[0]
+        self.assertEqual(test_list, ("item 1", "item 2"), "Failed to pair.")
+
+
+class InspectGetMembersTestCase(unittest.TestCase):
+
+    def test_inspect_get_members(self):
+        """ Test inspect get members """
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:610]: This one is a bit more complex. I would update this to:
`""" Test the builtin inspect.getmembers raises a TypeError for the dummy class, and that util.inspect_getmembers has the correct output."""`

File: src/odemis/util/test/util_test.py [hunk 8]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+        testing.start_backend(ENZEL_CONFIG)
         # check if the role now is enzel instead of sparc
         role = model.getMicroscope().role
         self.assertEqual(role, "enzel")
 
 
+class FindClosestTestCase(unittest.TestCase):
+
+    def test_simple(self):
+        """ Finding the closest point in a single list """
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:457]: It's not a hard rule but in general for one line docstrings we do not put spaces between the quotes and the text, so this one would become:
```
"""Finding the closest point in a single list"""
```

Feel free to leave it like you had it, or update it for all your test cases.

File: src/odemis/util/test/util_test.py [hunk 9]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+        in_exp = {
+            tuple([-1, -5, -6]): 0,
+            tuple([1, 2, 3]): 2,
+            tuple([0, 5, 10]): 1,
+            tuple([-15, 30, -60]): 0
+        }
+
+        for testlist, hardcoded_closest_index in in_exp.items():
+            closest = util.index_closest(5, testlist)
+            self.assertEqual(hardcoded_closest_index, closest,  "Failed to get correct output for "
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:496]: There's 2 spaces between `closest`, and `"Failed...`

File: src/odemis/util/test/util_test.py [hunk 10]
```diff
@@ -396,54 +401,255 @@ def test_line_clipping(self):
...
+                                                         "the closest value to 5 in %s" % in_exp)
+
+
+class IndexClosestTestCase(unittest.TestCase):
+
+    def test_simple(self):
+        """ Test finding the index of the closest point in a list """
+        simple_list = [1, 5, 6]
+        closest_index = util.index_closest(1, simple_list)
+        self.assertEqual(closest_index, 0,  "Failed to get correct output for "
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:482]: There's 2 spaces between `0`, and `"Failed...`

**Review by tepals — CHANGES_REQUESTED** [review]
The test cases already look good! Most my comments are on the style :).

**Review by tepals — APPROVED** [review]
